spinand.c

```diff
@@ -0,0 +1,279 @@
+
+	/* Send Read */
+	return spi_send_command(flash, 1 + addr_len + DUMMY_BYTE, len, cmd, bytes);
+}
+
+static int spi_nand_inpage_read(struct flashctx *flash, unsigned int row_addr, unsigned int column_addr,
+                                uint8_t *bytes, unsigned int len, unsigned int chunksize)
+{
+	/* Read data within one page */
+	int ret;
+
+	if (chunksize == 0)
+		chunksize = len;
+	ret = spi_nand_wait_for_page(flash, row_addr);
+	if (ret)
+		return ret;
+
+	while (len) {
+		unsigned int data_to_read = min(chunksize, len);
+		ret = spi_nand_read_cache(flash, column_addr, bytes, data_to_read);
+		if (ret)
+			return ret;
+		len -= data_to_read;
+		column_addr += data_to_read;
+		bytes += data_to_read;
+		printf("page %d\n",len);
+	}
+
+	return 0;
+}
+
+static inline unsigned int ilog2(unsigned int x) {
+	int res = 0;
+	while (x >>= 1)
+		++res;
+	return res;
+}
+
+int spi_nand_read_chunked(struct flashctx *flash, uint8_t *buf, unsigned int start,
+                          unsigned int len, unsigned int chunksize)
+{
+	/* Read data over multiple pages */
+	int ret;
+
+	/* Calculate acture column address len */
+	unsigned int column_addr_len = ilog2(flash->chip->page_size);
+	unsigned int column_addr_mask = ~(-1 << column_addr_len);
+	printf("%d %d %d\n",flash->chip->page_size,column_addr_len,column_addr_mask);
+
+	while (len) {
+		unsigned int data_to_read = min(chunksize, len);
+		ret = spi_nand_inpage_read(flash, start >> column_addr_len, start & column_addr_mask, buf, len, data_to_read);
+		if (ret)
+			return ret;
+		len -= data_to_read;
+		start += data_to_read;
+		buf += data_to_read;
+		printf("chunked %d\n",len);
+	}
+
+	return 0;
+}
+
+/***************************************
+ * PROGRAM CYCLE
+ ***************************************/
+
+
+/***************************************
+ * PROBE
+ ***************************************/
+
+static char *strncpy_tospace(char *dest, const char *src, unsigned int len)
+{
+	int i;
+
+	for (i = 0; i < len; i++) {
+		dest[i] = src[i];
+		if (dest[i] == ' ') {
+			dest[i] = '\0';
+			break;
+		}
+	}
+	return dest;
+}
+
+static void probe_spi_nand_toshiba(struct flashctx *flash, const uint8_t *parameters)
+{
+	char *vendor = (char *)malloc(12);
+	char *name = (char *)malloc(20);
+
+	flash->chip->vendor = strncpy_tospace(vendor, (const char *)(parameters + 32), sizeof(vendor));
+	flash->chip->name = strncpy_tospace(name, (const char *)(parameters + 44), sizeof(name));
+	flash->chip->page_size = *(uint32_t *)(parameters + 80);
+	flash->chip->data_per_spare = flash->chip->page_size / *(uint16_t *)(parameters + 84);
+	flash->chip->partial_page_size = *(uint32_t *)(parameters + 86);
+	flash->chip->block_size = flash->chip->page_size * *(uint32_t *)(parameters + 92);
+	flash->chip->unit_size = flash->chip->block_size * *(uint32_t *)(parameters + 96);
+	flash->chip->total_size = flash->chip->unit_size * parameters[100] / 1024;
+	flash->chip->total_size = 64;
```

Comment:
I think a github comment touches on this. They said they could only read 64k. IIRC this param is referenced in kilobytes.